src/webgpu/api/validation/resource_usages/textureUsageInRender.spec.ts

```diff
@@ -49,64 +55,173 @@ class TextureUsageTracking extends ValidationTest {
 
 export const g = makeTestGroup(TextureUsageTracking);
 
-const READ_BASE_LEVEL = 3;
-const READ_BASE_LAYER = 0;
-
-g.test('readwrite_upon_subresources')
-  .params([
-    // read and write usages are binding to the same texture subresource.
-    {
-      writeBaseLevel: READ_BASE_LEVEL,
-      writeBaseLayer: READ_BASE_LAYER,
-      _success: false,
-    },
-
-    // read and write usages are binding to different mip levels of the same texture.
-    {
-      writeBaseLevel: READ_BASE_LEVEL + 1,
-      writeBaseLayer: READ_BASE_LAYER,
-      _success: true,
-    },
-
-    // read and write usages are binding to different array layers of the same texture.
-    {
-      writeBaseLevel: READ_BASE_LEVEL,
-      writeBaseLayer: READ_BASE_LAYER + 1,
-      _success: true,
-    },
-  ])
-  .fn(async t => {
-    const { writeBaseLevel, writeBaseLayer, _success } = t.params;
+const BASE_LEVEL = 3;
+const BASE_LAYER = 0;
+const TOTAL_LEVELS = 6;
+const TOTAL_LAYERS = 2;
 
-    const texture = t.createTexture({ arrayLayerCount: 2, mipLevelCount: 6 });
+g.test('subresources_and_binding_types_combination_for_color')
+  .params(
+    params()
+      .combine([
+        // Two texture usages are binding to the same texture subresource.
+        {
+          baseLevel: BASE_LEVEL,
```

Comment:
@kainino0x I'm not sure how we want to do things in the CTS, maybe we could compute the success value in the test directly? Also maybe the list of ranges to test could be stored as a helper list and reused in several places? Finally it might be useful to test multiple ranges for both view0 and view1. (this comment doesn't need to be addressed in this PR)

Reply:
If we limit the computation to fulfill the requirement of this test, say only two views. It is not difficult. But if we need a common function to trace the resource overlap for various situations, it might be very complicated (like the impl in Dawn project). 

BTW, do you think it is necessary to test more cases about resource conflict, it is not difficult to add them, for example:
1) same layer: a few mip levels overlap or not with another few mip levels: layer = 0: view0_mip = [0, 1, 2], view1_mip = [3, 4] / view1_mip = [2, 3]. The current test we have only test view0 with one single subresource. 
2) same level: a few array layers overlap or not with another few layers
3) a few mip levels within the same layer overlap or not with a few array layers within the same level 
4) a few subresources (multiple levels * multiple layers) overlap or not with another few subresource (multiple levels * multiple layers).   

If you think it is necessary for cover resource subresource overlap, I can add it (maybe in another pr?). 

Looks like subresource overlap doesn't interact a lot with read-write usage conflict and their combinations might be unnecessary. But doing more tests in CTS is not a bad thing, per my understanding.

Reply:
Yeah it could be worth having a test that's more detailed in how the various resource ranges overlap, and this test can stay mostly as is.

Reply:
> I'm not sure how we want to do things in the CTS, maybe we could compute the success value in the test directly?

We have a lot of tests that go one way or the other. Usually we compute the success value in the test if the parameterizations are generated; otherwise it's generally not that hard to write them manually.

For some validation tests I've contemplated suggesting that we write a full TS reimplementation of the full validation rules from the spec, so you can throw arbitrary test cases at it and compare them with the browser implementation, but I'm never sure whether it's a good idea.

Agree it makes sense to have more detailed tests on resource range overlaps be separate from these.